### PR TITLE
refactor: untangle data flow structs

### PR DESCRIPTION
**Pull request by macklin-10x**
Works over the type hierarchy used to pass high-level data around, removing redundancy and tightening up mutation.

### CODE REVIEW

File: enclone_core/src/enclone_structs.rs [hunk 1]
```diff
@@ -60,8 +32,6 @@ pub struct EncloneExacts {
     pub join_info: Vec<JoinInfo>,
     pub drefs: Vec<DonorReferenceItem>,
     pub sr: Vec<Vec<Double>>,
-    pub fate: Vec<HashMap<String, BarcodeFate>>, // GETS MODIFIED SUBSEQUENTLY
```
Comment (macklin-10x) [inline enclone_core/src/enclone_structs.rs:63]: Not any more, it doesn't! :)

File: enclone_core/src/defs.rs [hunk 2]
```diff
@@ -249,6 +249,32 @@ pub struct GeneralOpt {
...
+impl GeneralOpt {
+    pub fn is_bcr(&self) -> bool {
+        if self.tcr {
+            assert!(!self.bcr);
+            return false;
+        }
+        // The original logic for computing this value was based on the confusing
+        // assumption that this is always true if self.tcr is false.
+        // I've preserved this for now, but this line might make sense to add:
+        // assert!(self.bcr);
```
Comment (sreenathkrishnan) [inline enclone_core/src/defs.rs:261]: Makes sense! We could use enum instead of two booleans

**Review by sreenathkrishnan — APPROVED** [review]
Nice simplification!

File: enclone_core/src/defs.rs (hunk 2 above)
Comment (macklin-10x) [reply enclone_core/src/defs.rs:261]: Yeah, the logic here was just so odd/implicit that I decided to back away slowly before turning and running

**Pull request merged**